sdk/armcore/internal/pollers/async/async.go

```diff
@@ -0,0 +1,118 @@
+// +build go1.13
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package async
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/armcore/internal/pollers"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+)
+
+const (
+	finalStateAsync = "azure-async-operation"
+	finalStateLoc   = "location"
+	finalStateOrig  = "original-uri"
+)
+
+// Applicable returns true if the LRO is using Azure-AsyncOperation.
+func Applicable(resp *azcore.Response) bool {
+	return resp.Header.Get(pollers.HeaderAzureAsync) != ""
+}
+
+// Poller is an LRO poller that uses the Azure-AsyncOperation pattern.
+type Poller struct {
+	Type       string `json:"type"`
+	AsyncURL   string `json:"asyncURL"`
+	LocURL     string `json:"locURL"`
+	OrigURL    string `json:"origURL"`
+	Method     string `json:"method"`
+	FinalState string `json:"finalState"`
+	CurState   string `json:"state"`
+}
+
+// New creates a new Poller from the provided initial response and final-state type.
+func New(resp *azcore.Response, finalState string, pollerID string) (*Poller, error) {
+	azcore.Log().Write(azcore.LogLongRunningOperation, "Using Azure-AsyncOperation poller.")
+	asyncURL := resp.Header.Get(pollers.HeaderAzureAsync)
+	if asyncURL == "" {
+		return nil, errors.New("response is missing Azure-AsyncOperation header")
+	}
+	if !pollers.IsValidURL(asyncURL) {
+		return nil, fmt.Errorf("invalid polling URL %s", asyncURL)
+	}
+	p := &Poller{
+		Type:       pollers.MakeID(pollerID, "async"),
+		AsyncURL:   asyncURL,
+		LocURL:     resp.Header.Get(pollers.HeaderLocation),
+		OrigURL:    resp.Request.URL.String(),
+		Method:     resp.Request.Method,
+		FinalState: finalState,
+	}
+	// check for provisioning state
+	state, err := pollers.GetProvisioningState(resp)
+	if errors.Is(err, pollers.ErrNoBody) || errors.Is(err, pollers.ErrNoProvisioningState) {
+		// NOTE: the ARM RPC spec explicitly states that for async PUT the initial response MUST
+		// contain a provisioning state.  to maintain compat with track 1 and other implementations
+		// we are explicitly relaxing this requirement.
+		/*if resp.Request.Method == http.MethodPut {
+			// initial response for a PUT requires a provisioning state
+			return nil, err
+		}*/
```

Comment:
Correct, we didn't enforce this in track 1 either (nor does python's implementation).  In fact, there's a test in the autorest LRO suite that covers this exact scenario.  I think it's likely to not be a problem in practice.  If we get here, it means we have an `Azure-AsyncOperation` header to poll for status on, so should be safe to assume the operation is in progress, we just make up the status message "InProgress" instead of what the endpoint should return.